gc/src/weak/weak_gc.rs

```diff
@@ -0,0 +1,264 @@
+pub use crate::gc::{finalizer_safe, force_collect};
+use crate::gc::{GcBox, GcBoxType};
+pub use crate::trace::{Finalize, Trace};
+use crate::weak::{clear_root_bit, Ephemeron};
+use crate::{set_data_ptr, GcPointer};
+use std::cell::Cell;
+use std::cmp::Ordering;
+use std::fmt::{self, Debug, Display};
+use std::hash::{Hash, Hasher};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::Deref;
+use std::ptr::NonNull;
+use std::rc::Rc;
+
+////////////
+// WeakGc //
+////////////
+
+/// A weak Garbage Collected pointer for an immutable value
+///
+/// This implementation uses an Ephemeron as a generalized weak
+/// box to trace and sweep the inner values.
+pub struct WeakGc<T: Trace + ?Sized + 'static> {
+    ptr_root: Cell<NonNull<GcBox<Ephemeron<T>>>>,
+    marker: PhantomData<Rc<T>>,
+}
```

Comment:
The extra pointer indirection shouldn’t be necessary here. We should be able to simply use

```rust
pub struct WeakGc<T: Trace + ?Sized + 'static> {
    ephemeron: Ephemeron<T, ()>,
}
```

right?

Reply:
I believe we need the `GcBox` pointer that contains the `Ephemeron` here. The `weak_trace` that runs during the ephemeron marking phase needs to be able to enqueue the pointer...am I overthinking it?